### PR TITLE
Added hw solution old topic security

### PR DESCRIPTION
**Pull request by vitaliiplys**

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 1]
```diff
@@ -27,31 +28,36 @@
 public class BookController {
     private final BookService bookService;
 
+    @PreAuthorize("hasAnyRole('ROLE_USER','ROLE_ADMIN')")
```
Comment (okuzan) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java]: you can omit this with no change in functionality, you secured it in SecurityConfig - every endpoint apart from authentication requires some role, so it's enough to specify methods that require only ADMIN

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    private String firstName;
+
+    @Column(name = "last_name",nullable = false)
+    private String lastName;
+
+    @Column(name = "shipping_address")
+    private String shippingAddress;
+
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(name = "users_roles",
```
Comment (okuzan) [inline src/main/java/com/example/onlinebookstore/model/User.java]:
```suggestion
    @JoinTable(name = "user_role",
```
IMO it's concise and clear

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 3]
```diff
@@ -1,3 +1,13 @@
...
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
+  - include:
+      file: db/changelog/changes/03-create-roles-table.yaml
+  - include:
+      file: db/changelog/changes/04-create-users-roles-table.yaml
+  - include:
+      file: db/changelog/changes/05-insert-roles-table.yaml
+  - include:
+      file: db/changelog/changes/06-insert-users-table.yaml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: make sure each your file ends with one empty line

File: src/main/java/com/example/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
+            throw new RegistrationException("Unable to complete registration.");
+        }
+        User user = new User();
+        user.setEmail(request.getEmail());
+        user.setPassword(passwordEncoder.encode(request.getPassword()));
+        user.setFirstName(request.getFirstName());
+        user.setLastName(request.getLastName());
+        user.setShippingAddress(request.getShippingAddress());
```
Comment (okuzan) [inline src/main/java/com/example/onlinebookstore/service/user/impl/UserServiceImpl.java:36]: use usermapper for this

File: src/main/java/com/example/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        Role role = roleRepository.findByName(Role.RoleName.ROLE_USER).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find role" + Role.RoleName.ROLE_USER)
+        );
```
Comment (okuzan) [inline src/main/java/com/example/onlinebookstore/service/user/impl/UserServiceImpl.java]: let's refactor it, move this logic to a separate service

        Role userRole = roleService.getRoleByRoleName(Role.RoleName.ROLE_USER);
        user.setRoles(new HashSet<>(Set.of(userRole)));

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job overall

**Review by kozhukhovsky — APPROVED** [review]

**Pull request merged**